File: ToBeTranslated/Lec6_forClass.R


#############################################
#      LECTURE 6: Data analysis             #
#############################################

# We will first go through (almost) the same steps as in 
# Lecture 5. This time, we will organize the code in a more
# efficient way, such that it becomes more "reusable"
# You will also learn how to merge two data sets.
# The most important thing today are scatter plots
# and regression analysis. These are key tools for
# data analysis.

rm(list = ls()) # Empty workspace to start with a "clean sheet"

# REPLACE THE WORKING DIRECTORY BELOW WITH THE ONE FOR YOUR DEVICE
setwd("/Users/dominiquepaul/xJob/DataWithPythonCourse")
#load("rawExpImp (1).RData")

# Read files, adjust file names to your situation!

rawXrates = read.csv(file = 
   "dataXrates.csv", sep = ";")

rawAussen = read.csv(file = 
    "aussenhandel_snb.csv", sep = ";")

# If you are having troubles with the data, use the following

# load("rawXrates.RData")
# load("rawExpImp.RData")

# You get these data from the online script on
# https://binswanger.github.io/practicaldata_hs16/#Problems--What-if-some-numbers-are-automatically-converted-to-dates-
# Put these files in your working directory

# SET PARAMETERS HERE!
######################

startYear = 2000
curr = "EUR1"
tradeDir = "A"  # Direction of trade
                # Values are E (Einfuhr) 
                # and A (Ausfuhr)
                # run unique(rawAussen$D0) for overview
goodsType = "MAE" # run unique(rawAussen$D1)

measure = "R" # run unique(rawAussen$D2)

a = rawXrates[1:20,]
a
head(a,-1)
tail(a,-1)

unique(rawAussen$D0)
unique(rawAussen$D1)
unique(rawAussen$D2)

# Functions
###########


toGrowth = function(x){
  out = (tail(x, -1)/head(x, -1) -1 )*100
  # This is one element shorter than input, 
  # add an NA as first value
  out = c(NA, out)
  return(out)
}




# Customize the data
####################

# A unique time identifier for BOTH data
library(stringr)

rawXrates$year = as.numeric(  
  substr(rawXrates$Date, start = 1, stop = 4)  )

rawXrates$month = as.numeric(  
  substr(rawXrates$Date, start = 6, stop = 7)  )

rawXrates$timeID = rawXrates$year + 
  (rawXrates$month-1)/12


rawAussen$year = as.numeric(  
  substr(rawAussen$Date, start = 1, stop = 4)  )

rawAussen$month = as.numeric(  
  substr(rawAussen$Date, start = 6, stop = 7)  )

rawAussen$timeID = rawAussen$year + 
  (rawAussen$month-1)/12



xrates = subset(rawXrates, 
        D1 == curr &  # NOTE the apparance of the VARIABLE xrate!!!
           D0 == "M0" &
          timeID >= startYear , # NOTE the apparance of the VARIABLE startYear!!!
        select = c("timeID", "D1", "Value"))

aussen = 
  subset(rawAussen, 
         
         D0 == tradeDir  &  
         D1 == goodsType &
         D2 == measure   &
         timeID >= startYear , 
         
         select = c("timeID", "D0","D1","D2", "Value"))


# Bring data into wide format
library(reshape2)
xrates1= dcast(xrates, timeID ~ D1, value.var = "Value")
xrates
xrates1

aussen1 = dcast(aussen, timeID ~ D0 + D1 + D2, value.var = "Value")
aussen
aussen1
# Merge the two data sets (NEW!!!!)
# DA stands for "Data for Analysis"
DA = merge(xrates, aussen, by = "timeID")
# The "by" argument contains the so-called "key".


# [FOR LATER] convert to growth rates
DA[[curr]] = toGrowth(DA[[curr]])


# ANALYSIS
###########

# Variable name for exports/imports
names(DA)
vn = paste(tradeDir, goodsType, measure, sep = "_")



# Make a plot
plot(DA[[curr]], DA[[vn]], 
        # NOTE: This code works for all types of export/import
        # data and exchange rates!!! They are captured in the 
        # variable names
        
        pch = 16, # data points as dots; google "r plot pch"
     
        cex = .7, # the size of the point (exam question from last semester :-)
     
        xlab = curr, ylab = vn, # axis labels
     
        col = c("turquoise") # color of dots
     
     )  # The closing paranthesis

# Add a regression line

abline(  lm(DA[[vn]]  ~   DA[[curr]]
          ), col="red3", lwd = 3) 



# Run a regression
reg = lm(DA[[vn]]~DA[[curr]])
summary(reg)





# HOWEVER... It is more interesting to 
# look at how CHANGES in exchanges rates
# affect CHANGES in exports/imports

# How calculate changes?

test = 1:10
head(test, -1)
tail(test, -1)

dTest = (tail(test, -1)/head(test, -1) -1)*100

File: Lec6_forClass.py

import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf

# Set the working directory to the folder
# where you have the csv files from the SNB

os.chdir("/Users/dominiquepaul/xJob/DataWithPythonCourse")
# os.chdir("YOUR DIRECTORY")

# you might have to change the separator
rawXrates = pd.read_csv('dataXrates1.csv', sep = ",")
rawAussen = pd.read_csv('aussenhandel_snb1.csv', sep = ",")

# we convert our non-numeric values to numbers
rawXrates.Value = pd.to_numeric(rawXrates.Value)
rawAussen.Value = pd.to_numeric(rawAussen.Value)



# SET PARAMETERS HERE!
######################

startYear = 2000

curr = "EUR1"
# the currency examined, you can run
# rawXrates.D1.unique()
# to check for all possible values

tradeDir = "A"  
# Direction of trade               
 # Values are E (Einfuhr)                 
 # and A (Ausfuhr)                
 #run rawAussen.D0.unique() for overview
 
goodsType = "MAE" 
# run rawAussen.D1.unique()

measure = "R" 
#run rawAussen.D2.unique()



a = rawXrates.iloc[0:20, :]
a
a.head(-1)
a.tail(-1)


rawAussen.D0.unique()
rawAussen.D1.unique()
rawAussen.D2.unique()


# Functions
###########


def to_growth(series):
    #we have to reset the index as pandas will otherwise divide
    # all rows which have the same index, which would mean that all 
    # results would be zero
    out = (series.tail(-1).reset_index(drop = True) / series.head(-1) -1 ).reset_index(drop = True) * 100
    # This is one element shorter than input, 
    # add an NA as first value
    df1 = pd.DataFrame([None])
    # 'ignore_index' means that the two DFs are merged while neglecting 
    # any order of their indices 
    out = pd.concat([df1, out], ignore_index = True)
    return out

# Customize the data
####################

# A unique time identifier for BOTH data

rawXrates["year"] = pd.to_numeric(rawXrates["Date"].str[0:4])
rawXrates["month"] = pd.to_numeric(rawXrates["Date"].str[5:7])
rawXrates["timeID"] = rawXrates.year + (rawXrates.month -1) / 12

rawAussen["year"] = pd.to_numeric(rawAussen["Date"].str[0:4])
rawAussen["month"] = pd.to_numeric(rawAussen["Date"].str[5:7])
rawAussen["timeID"] = rawAussen.year + (rawAussen.month -1) / 12

# we subset our data

xrates = rawXrates[(rawXrates.D1 == curr) & (rawXrates.D0 == "M0") & 
          (rawXrates.timeID >= startYear)].loc[:, ["timeID", "D1", "Value"]]           

aussen = rawAussen[(rawAussen.D0 == tradeDir) & (rawAussen.D1 == goodsType) & 
                   (rawAussen.D2 == measure) & (rawAussen.timeID >= 
                   startYear)].loc[:, ["timeID", "D0", "D1", "D2", "Value"]]

                   
# Bring data into wide format
xrates_wide = xrates.pivot(index = "timeID", columns = "D1",  values = "Value")
                   
aussen_wide = pd.pivot_table(aussen, index='timeID', columns=['D0', 'D1', 'D2'], aggfunc=max)             
  


##################################
# Merge the two data sets (NEW!!!!)
# DA stands for "Data for Analysis"

# both tables are joined together by their index which currently is 
# the timeID, look at the data and you will see that timeID is NOT a column
# but the index!

# if you only want to join rows where values for both DFs are present you can 
# use 'join = inner'
DA = pd.concat([xrates_wide, aussen_wide], axis=1)
# we reset our index, 'drop = False' pushes the index into the df as a column
DA = DA.reset_index(drop = False)

# [FOR LATER] convert to growth rates
DA[curr] = to_growth(DA.loc[:, curr])




###########
# ANALYSIS
###########


# Variable name for exports/imports
DA.columns
trade_col = tradeDir + "_" + goodsType + "_" + measure
# rename column names
DA.columns = ["TimeID", curr, trade_col]

DA

# for convenience we define: 
x = DA[curr]
y = DA[trade_col]

# Make a plot
plt.plot(x, 
         y, 
         # NOTE: This code works for all types of export/import
         # data and exchange rates!!! They are captured in the 
         # variable names
         
         'b.', #color and form of the dots: 'b' => blue, '.' => dots
         ms = 5 # point size
         )  #The closing paranthesis

# labels x-axis
plt.xlabel(curr)
# labels y-axis
plt.ylabel(trade_col)
plt.show


# add a regression line 

# this checks which rows contain values that are and are not NaN values 
# because dont want to use these when calculating our OLS line
idx = np.isfinite(x) & np.isfinite(y)

# find the line of best fit with np.polyfit
m,c = np.polyfit(x[idx], y[idx], 1)

# we plot the line:
plt.plot(x, 
         m*x + c,   # our y-values in the form of a linear equation
         'r-',      # a solid red line 'r' => red; '-' => solid line
         lw = 0.8   # set the line width
         ) 



# run a regression
# we again omit all rows containing NaN
results = smf.OLS(y[idx], x[idx]).fit()

# Inspect the results
print(results.summary())





# HOWEVER... It is more interesting to 
# look at how CHANGES in exchanges rates
# affect CHANGES in exports/imports

# How calculate changes?

test = pd.Series(range(1,11))
test.head(-1)
test.tail(-1)

d_test = (test.tail(-1).reset_index(drop = True)/test.head(-1).reset_index(drop = True) -1)*100
d_test

# you can test what would happen if we wouldnt use 'reset.index(drop = True)'
d_test_2 = (test.tail(-1) / test.head(-1) -1)*100           
d_test_2              